src/VisualStudio/Core/Def/DesignerAttribute/VisualStudioDesignerAttributeService.cs

```diff
@@ -52,12 +52,32 @@ internal class VisualStudioDesignerAttributeService :
         /// </summary>
         private IVSMDDesignerService? _legacyDesignerService;
 
+        /// <summary>
+        /// Queue that tells us to recompute designer attributes when workspace changes happen.  This queue is
+        /// cancellable and will be restarted when new changes come in.  That way we can be quickly update the designer
+        /// attribute for a file when a user edits it.
+        /// </summary>
         private readonly AsyncBatchingWorkQueue _workQueue;
 
-        // We'll get notifications from the OOP server about new attribute arguments. Collect those notifications and
-        // deliver them to VS in batches to prevent flooding the UI thread.
+        /// <summary>
+        /// We'll get notifications from the OOP server about new attribute arguments. Collect those notifications and
+        /// deliver them to VS in batches to prevent flooding the UI thread.  Importantly, we do not cancel this queue.
+        /// Once we've decided to update the project system, we want to allow that to proceed.
+        /// </summary>
         private readonly AsyncBatchingWorkQueue<DesignerAttributeData> _projectSystemNotificationQueue;
 
+        /// <summary>
+        /// Keep track of the last version we were at when we processed a project.  We'll skip reprocessing projects if
+        /// that version hasn't changed. don't change.
+        /// </summary>
+        private readonly ConcurrentDictionary<ProjectId, VersionStamp> _projectToLastComputedDependentSemanticVersion = new();
+
+        /// <summary>
+        /// Keep track of the last information we reported per document.  We will avoid notifying the host if we
+        /// recompute and these don't change.
+        /// </summary>
+        private readonly ConcurrentDictionary<DocumentId, DesignerAttributeData> _documentToLastReportedData = new();
```

Comment:
here is the state teh host side keeps to know if it needs to call into the remote side, and to know when it needs to notify the project system about actually changed data.